src/main/kotlin/org/rust/lang/core/psi/ext/RsQualifiedNamedElement.kt

```diff
@@ -356,29 +356,37 @@ data class RsQualifiedName private constructor(
         }
 
         private fun RsQualifiedNamedElement.toParentItem(): Item? {
-            val name = ((this as? RsMod)?.modName ?: name) ?: return null
+            val name = when {
+                this is RsFunction && isProcMacroDef -> procMacroName
+                this is RsMod && isCrateRoot -> containingCrate?.normName
+                this is RsMod -> modName
+                else -> name
+            } ?: return null
+
             val itemType = when (this) {
                 is RsStructItem -> when (kind) {
                     RsStructKind.STRUCT -> STRUCT
                     RsStructKind.UNION -> UNION
                 }
+
                 is RsEnumItem -> ENUM
                 is RsTraitItem -> TRAIT
                 is RsTraitAlias -> TRAITALIAS
                 is RsTypeAlias -> TYPE
-                is RsFunction -> FN
+                is RsFunction -> when {
+                    isCustomDeriveProcMacroDef -> DERIVE
+                    isAttributeProcMacroDef -> ATTR
+                    isBangProcMacroDef -> MACRO
+                    else -> FN
+                }
```

Comment:
Lets optimize fast path a bit

```suggestion
                is RsFunction -> when {
                    !isProcMacroDef -> FN
                    isCustomDeriveProcMacroDef -> DERIVE
                    isAttributeProcMacroDef -> ATTR
                    isBangProcMacroDef -> MACRO
                    else -> return null
                }
```